File: bosun/jobs.py
import hashlib
import re
from kubernetes import client, config
from . import config as settings
from .logs import get_logger

log = get_logger("bosun.jobs")

MAX_NAME = 63


class JobExists(Exception):
    """A Job for this delivery already exists (GitHub retried the webhook)."""


class AtCapacity(Exception):
    """Too many reviews are already running."""


def active_reviews(batch=None) -> int:
    """Count reviewer Jobs that are neither complete nor failed."""
    batch = batch or client.BatchV1Api()
    jobs = batch.list_namespaced_job(
        settings.NAMESPACE, label_selector="app.kubernetes.io/managed-by=bosun"
    )
    running = 0
    for job in jobs.items:
        status = job.status
        if status is None:
            running += 1
            continue
        finished = (status.succeeded or 0) or (status.failed or 0)
        if not finished:
            running += 1
    return running


def _name(repo: str, ref: str, delivery: str) -> str:
    """Build a DNS-safe, collision-free Job name.

    The delivery suffix is appended after truncation so that long branch names
    cannot push it off the end and make two different deliveries collide.
    """
    suffix = hashlib.sha256(f"{repo}\n{ref}\n{delivery}".encode()).hexdigest()[:10]
    raw = f"review-{repo.split('/')[-1]}-{ref}".lower()
    stem = re.sub(r"[^a-z0-9-]+", "-", raw).strip("-")
    stem = stem[: MAX_NAME - len(suffix) - 1].strip("-")
    return f"{stem}-{suffix}" if stem else f"review-{suffix}"


def submit_review(repo: str, ref: str, sha: str, delivery: str, pr_number: int | None, trigger: str) -> str:
    try:
        config.load_incluster_config()
    except config.ConfigException:
        config.load_kube_config()

    batch = client.BatchV1Api()
    running = active_reviews(batch)
    if running >= settings.MAX_CONCURRENT_REVIEWS:
        log.warning(
            "rejecting review: at capacity",
            extra={"context": {"repo": repo, "running": running,
                               "cap": settings.MAX_CONCURRENT_REVIEWS, "delivery": delivery}},
        )
        raise AtCapacity(f"{running} reviews already running (cap {settings.MAX_CONCURRENT_REVIEWS})")

    name = _name(repo, ref, delivery)
    env = [
        client.V1EnvVar(name="BOSUN_REPO", value=repo),
        client.V1EnvVar(name="BOSUN_REF", value=ref),
        client.V1EnvVar(name="BOSUN_SHA", value=sha),
        client.V1EnvVar(name="BOSUN_TRIGGER", value=trigger),
        client.V1EnvVar(name="BOSUN_REVIEW_PROVIDER", value=settings.REVIEW_PROVIDER),
        client.V1EnvVar(name="BOSUN_REVIEW_TIMEOUT_SECONDS", value=str(settings.REVIEW_TIMEOUT_SECONDS)),
        client.V1EnvVar(name="BOSUN_PR_NUMBER", value=str(pr_number or "")),
        client.V1EnvVar(name="GITHUB_TOKEN", value_from=client.V1EnvVarSource(secret_key_ref=client.V1SecretKeySelector(name=settings.GITHUB_TOKEN_SECRET, key="token", optional=True))),
        client.V1EnvVar(name="OPENAI_API_KEY", value_from=client.V1EnvVarSource(secret_key_ref=client.V1SecretKeySelector(name=settings.AI_SECRET, key="openai-api-key", optional=True))),
        client.V1EnvVar(name="CLAUDE_CODE_OAUTH_TOKEN", value_from=client.V1EnvVarSource(secret_key_ref=client.V1SecretKeySelector(name=settings.AI_SECRET, key="claude-code-oauth-token", optional=True))),
        client.V1EnvVar(name="ANTHROPIC_API_KEY", value_from=client.V1EnvVarSource(secret_key_ref=client.V1SecretKeySelector(name=settings.AI_SECRET, key="anthropic-api-key", optional=True))),
        client.V1EnvVar(name="GEMINI_API_KEY", value_from=client.V1EnvVarSource(secret_key_ref=client.V1SecretKeySelector(name=settings.AI_SECRET, key="gemini-api-key", optional=True))),
        # File-based provider credentials; the reviewer writes these to disk.
        client.V1EnvVar(name="CODEX_AUTH", value_from=client.V1EnvVarSource(secret_key_ref=client.V1SecretKeySelector(name=settings.AI_SECRET, key="codex-auth", optional=True))),
        client.V1EnvVar(name="CLAUDE_CREDENTIALS", value_from=client.V1EnvVarSource(secret_key_ref=client.V1SecretKeySelector(name=settings.AI_SECRET, key="claude-credentials", optional=True))),
        client.V1EnvVar(name="BOSUN_GITHUB_APP_ID", value_from=client.V1EnvVarSource(secret_key_ref=client.V1SecretKeySelector(name=settings.GITHUB_TOKEN_SECRET, key="app-id", optional=True))),
        client.V1EnvVar(name="BOSUN_GITHUB_PRIVATE_KEY", value_from=client.V1EnvVarSource(secret_key_ref=client.V1SecretKeySelector(name=settings.GITHUB_TOKEN_SECRET, key="private-key", optional=True))),
        client.V1EnvVar(name="BOSUN_LOG_FORMAT", value=settings.LOG_FORMAT),
        client.V1EnvVar(name="BOSUN_LOG_LEVEL", value=settings.LOG_LEVEL),
    ]
    container = client.V1Container(
        name="reviewer", image=settings.REVIEW_IMAGE,
        command=["python3", "-m", "bosun.worker"], env=env,
        resources=client.V1ResourceRequirements(
            requests={"cpu": "250m", "memory": "512Mi"},
            limits={"cpu": "2", "memory": "2Gi"},
        ),
        security_context=client.V1SecurityContext(
            allow_privilege_escalation=False,
            run_as_non_root=True,
            run_as_user=settings.RUN_AS_USER,
            capabilities=client.V1Capabilities(drop=["ALL"]),
        ),
    )
    pod = client.V1PodTemplateSpec(
        metadata=client.V1ObjectMeta(labels={"app.kubernetes.io/name": "bosun-review", "bosun/repository": repo.replace("/", "-")[:63]}),
        spec=client.V1PodSpec(restart_policy="Never", containers=[container], automount_service_account_token=False),
    )
    job = client.V1Job(
        metadata=client.V1ObjectMeta(name=name, labels={"app.kubernetes.io/managed-by": "bosun"}),
        spec=client.V1JobSpec(
            template=pod, backoff_limit=0,
            ttl_seconds_after_finished=settings.JOB_TTL_SECONDS,
            active_deadline_seconds=settings.JOB_DEADLINE_SECONDS,
        ),
    )
    try:
        batch.create_namespaced_job(settings.NAMESPACE, job)
    except client.ApiException as exc:
        if exc.status == 409:
            raise JobExists(name) from exc
        raise
    log.info(
        "review job created",
        extra={"context": {"job": name, "repo": repo, "ref": ref, "trigger": trigger,
                           "pr": pr_number or "", "running": running + 1}},
    )
    return name

File: internal/jobs/jobs.go
package jobs

import (
	"context"
	"crypto/sha256"
	"fmt"
	"regexp"
	"strings"

	"github.com/everydaydevopsio/bosun/internal/config"
	"github.com/everydaydevopsio/bosun/internal/review"
	batchv1 "k8s.io/api/batch/v1"
	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
)

const MaxName = 63

var unsafe = regexp.MustCompile(`[^a-z0-9-]+`)

type ExistsError struct{ Name string }

func (e ExistsError) Error() string { return "review job already exists: " + e.Name }
func (e ExistsError) Duplicate()    {}

type CapacityError struct{ Running, Cap int }

func (e CapacityError) Error() string {
	return fmt.Sprintf("%d reviews already running (cap %d)", e.Running, e.Cap)
}
func (e CapacityError) Capacity() {}

func Name(repo, ref, delivery string) string {
	sum := sha256.Sum256([]byte(repo + "\n" + ref + "\n" + delivery))
	suffix := fmt.Sprintf("%x", sum[:])[:10]
	stem := unsafe.ReplaceAllString(strings.ToLower("review-"+last(repo)+"-"+ref), "-")
	stem = strings.Trim(stem, "-")
	limit := MaxName - len(suffix) - 1
	if len(stem) > limit {
		stem = strings.Trim(stem[:limit], "-")
	}
	if stem == "" {
		return "review-" + suffix
	}
	return stem + "-" + suffix
}
func last(s string) string {
	if i := strings.LastIndex(s, "/"); i >= 0 {
		return s[i+1:]
	}
	return s
}

func Active(ctx context.Context, client kubernetes.Interface, namespace string) (int, error) {
	list, err := client.BatchV1().Jobs(namespace).List(ctx, metav1.ListOptions{LabelSelector: "app.kubernetes.io/managed-by=bosun"})
	if err != nil {
		return 0, err
	}
	n := 0
	for _, j := range list.Items {
		if j.Status.Succeeded == 0 && j.Status.Failed == 0 {
			n++
		}
	}
	return n, nil
}
func Submit(ctx context.Context, client kubernetes.Interface, cfg config.Config, req review.Request, delivery string) (string, error) {
	n, err := Active(ctx, client, cfg.Namespace)
	if err != nil {
		return "", err
	}
	if n >= cfg.MaxConcurrentReviews {
		return "", CapacityError{n, cfg.MaxConcurrentReviews}
	}
	name := Name(req.Repo, req.Ref, delivery)
	job := build(cfg, req, name)
	_, err = client.BatchV1().Jobs(cfg.Namespace).Create(ctx, job, metav1.CreateOptions{})
	if errors.IsAlreadyExists(err) {
		return "", ExistsError{name}
	}
	if err != nil {
		return "", err
	}
	return name, nil
}
func ptr[T any](v T) *T { return &v }
func env(cfg config.Config, req review.Request) []corev1.EnvVar {
	base := []corev1.EnvVar{{Name: "BOSUN_REPO", Value: req.Repo}, {Name: "BOSUN_REF", Value: req.Ref}, {Name: "BOSUN_SHA", Value: req.SHA}, {Name: "BOSUN_TRIGGER", Value: req.Trigger}, {Name: "BOSUN_REVIEW_PROVIDER", Value: cfg.ReviewProvider}, {Name: "BOSUN_REVIEW_TIMEOUT_SECONDS", Value: fmt.Sprint(cfg.ReviewTimeoutSeconds)}, {Name: "BOSUN_PR_NUMBER", Value: fmt.Sprint(req.PRNumber)}}
	for _, x := range []struct{ n, s, k string }{{"GITHUB_TOKEN", cfg.GitHubTokenSecret, "token"}, {"OPENAI_API_KEY", cfg.AISecret, "openai-api-key"}, {"CLAUDE_CODE_OAUTH_TOKEN", cfg.AISecret, "claude-code-oauth-token"}, {"ANTHROPIC_API_KEY", cfg.AISecret, "anthropic-api-key"}, {"GEMINI_API_KEY", cfg.AISecret, "gemini-api-key"}, {"CODEX_AUTH", cfg.AISecret, "codex-auth"}, {"CLAUDE_CREDENTIALS", cfg.AISecret, "claude-credentials"}, {"BOSUN_GITHUB_APP_ID", cfg.GitHubTokenSecret, "app-id"}, {"BOSUN_GITHUB_PRIVATE_KEY", cfg.GitHubTokenSecret, "private-key"}} {
		base = append(base, corev1.EnvVar{Name: x.n, ValueFrom: &corev1.EnvVarSource{SecretKeyRef: &corev1.SecretKeySelector{LocalObjectReference: corev1.LocalObjectReference{Name: x.s}, Key: x.k, Optional: ptr(true)}}})
	}
	return base
}
func build(cfg config.Config, req review.Request, name string) *batchv1.Job {
	deadline := cfg.ReviewTimeoutSeconds + 120
	return &batchv1.Job{
		ObjectMeta: metav1.ObjectMeta{Name: name, Labels: map[string]string{"app.kubernetes.io/managed-by": "bosun"}},
		Spec: batchv1.JobSpec{
			BackoffLimit: ptr(int32(0)), TTLSecondsAfterFinished: &cfg.JobTTLSeconds, ActiveDeadlineSeconds: &deadline,
			Template: corev1.PodTemplateSpec{
				ObjectMeta: metav1.ObjectMeta{Labels: map[string]string{"app.kubernetes.io/name": "bosun-review", "bosun/repository": strings.ReplaceAll(req.Repo, "/", "-")}},
				Spec: corev1.PodSpec{RestartPolicy: corev1.RestartPolicyNever, AutomountServiceAccountToken: ptr(false), Containers: []corev1.Container{{
					Name: "reviewer", Image: cfg.ReviewImage, Command: []string{"/usr/local/bin/bosun", "reviewer"}, Env: env(cfg, req),
					SecurityContext: &corev1.SecurityContext{AllowPrivilegeEscalation: ptr(false), RunAsNonRoot: ptr(true), RunAsUser: &cfg.RunAsUser, Capabilities: &corev1.Capabilities{Drop: []corev1.Capability{"ALL"}}},
				}}},
			},
		},
	}
}
